### PR TITLE
PageFormat QC test case generators

### PR DESCRIPTION
**Pull request by dcoutts**
Replace the PageLogical test case type with several new ones.

The new generators have more carefully thought out and measured size distributions.

In the tests for PageFormat itself, we can now cover multiple disk page sizes, not just 4k.

For the real vs reference tests we can now be more precise about when we want key/op sequences that must fit within a (4k) page, or could be bigger, must be sorted or not etc.

So far this just removes the old type and does a basic replacement with the new. It doesn't yet fully take advantage of the new test cases and generators by simplifying existing tests or adding new ones. That's to follow.

### CODE REVIEW

File: src-extras/Database/LSMTree/Extras/ReferenceImpl.hs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    shrink (PageContentOrdered kops) =
+      map PageContentOrdered (shrinkOrderedKeyOps kops)
+
+instance Arbitrary PageContentMaybeOverfull where
+    arbitrary =
+      PageContentMaybeOverfull <$>
+        genPageContentMaybeOverfull DiskPage4k noMinKeySize
+
+    shrink (PageContentMaybeOverfull kops) =
+      map PageContentMaybeOverfull (shrinkOrderedKeyOps kops)
```
Comment (jorisdral) [inline src-extras/Database/LSMTree/Extras/ReferenceImpl.hs:97]: Is it important for that this shrinks as if the key/ops were ordered?

File: test/Test/Database/LSMTree/Generators.hs [hunk 2]
```diff
@@ -26,14 +26,15 @@ import           Database.LSMTree.Internal.Serialise
...
-import           Test.Tasty (TestTree, testGroup)
-import           Test.Tasty.QuickCheck (testProperty)
+import           Test.Tasty (TestTree, localOption, testGroup)
+import           Test.Tasty.QuickCheck (QuickCheckMaxSize (..), testProperty)
 
 tests :: TestTree
 tests = testGroup "Test.Database.LSMTree.Generators" [
       testGroup "Range-finder bit-precision" $
         prop_arbitraryAndShrinkPreserveInvariant rfprecInvariant
-    , testGroup "LogicalPageSummaries" $
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Generators.hs:36]: :+1:

File: src-extras/Database/LSMTree/Extras/ReferenceImpl.hs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    orderdKeyOps,
+    shrinkKeyOps,
+    shrinkOrderedKeyOps,
+) where
+
+import           FormatPage
+import           Test.QuickCheck
+
+-- | A test case consisting of a key\/operation sequence that is guaranteed to
+-- fit into a 4k disk page.
```
Comment (jorisdral) [inline src-extras/Database/LSMTree/Extras/ReferenceImpl.hs:44]: I think this is still slightly misleading terminology: it allows for overflow

File: prototypes/FormatPage.hs [hunk 4]
```diff
@@ -503,23 +509,264 @@ fromIntegralChecked x
...
+  where
+    genkey = genKeyMinSize dpgsz minkeysz
+    genval = arbitrary
+
+-- | This generates a reasonable \"middle\" distribution of page sizes
+-- (relative to the given disk page size). In particular it covers:
+
+-- * small pages (~45% for 4k pages, ~15% for 64k pages)
+-- * near-maximum pages (~20% for 4k pages, ~20% for 64k pages)
+-- * some in between (~35% for 4k pages, ~60% for 64k pages)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply prototypes/FormatPage.hs:611]: It means that for 4k page sizes, the fraction of small, near-max and in-between is 45%, 20% and 35% respectively.

And similarly for 64k page sizes, but with different fractions.

File: prototypes/FormatPage.hs [hunk 5]
```diff
@@ -582,21 +826,68 @@ mkPageLogical =
...
+
+      , testProperty "genPageContentMaybeOverfull" $
+        checkCoverage $
+        forAll (genPageContentMaybeOverfull dpgsz noMinKeySize) $
+          prop_size_distribution dpgsz
+            (cover 10 True "over-full" (property True))
+
+      , testProperty "genPageContentSingle" $
+        checkCoverage $
+        coverTable "page size in disk pages"
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply prototypes/FormatPage.hs:834]: Sigh. Yes. Suggestions welcome.

File: prototypes/FormatPage.hs [hunk 6]
```diff
@@ -503,23 +509,264 @@ fromIntegralChecked x
...
+-- * nearly full pages
+-- * plus single key pages (possibly using one or more overflow pages)
+-- * a corner case of a single large key\/operation pair followed by some small
+--   key op pairs.
+--
+-- The keys are /not/ ordered: use 'orderdKeyOps' to sort and de-duplicate them
+-- if that is needed (but note this will change the order of key sizes).
+--
+genPageContentFits :: DiskPageSize -> MinKeySize -> Gen [(Key, Operation)]
+genPageContentFits dpgsz minkeysz =
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply prototypes/FormatPage.hs:558]: No need to trust my judgement. See the distribution in the test output for `testProperty "genPageContentFits"`.

File: src-extras/Database/LSMTree/Extras/ReferenceImpl.hs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    encodePage,
+    decodePage,
+    serialisePage,
+    deserialisePage,
+
+    -- * Test case types and generators
+    PageContentFits(..),
+    PageContentOrdered(..),
+    PageContentMaybeOverfull(..),
+    PageContentSingle(..),
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src-extras/Database/LSMTree/Extras/ReferenceImpl.hs:26]: Yes I was thinking that too, for the invariants. The tests of the distribution are covered already in the tests for `FormatPage` (and checking that the ones that are supposed to fit do indeed fit).

File: bench/micro/Bench/Database/LSMTree/Internal/RawPage.hs [hunk 8]
```diff
@@ -28,12 +29,12 @@ benchmarks = rawpage `deepseq` bgroup "Bench.Database.LSMTree.Internal.RawPage"
     , bench "existing-last" $ whnf (rawPageLookup rawpage) existingLast
     ]
   where
-    page :: PageLogical
-    page = unGen (genFullPageLogical DiskPage4k genSmallKey genSmallValue)
+    kops :: [(Key, Operation)]
+    kops = unGen (genPageContentNearFull DiskPage4k genSmallKey genSmallValue)
                  (mkQCGen 42) 200
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply bench/micro/Bench/Database/LSMTree/Internal/RawPage.hs:34]: Good catch, you're quite right.

File: src-extras/Database/LSMTree/Extras/ReferenceImpl.hs (hunk 1 above)
Comment (dcoutts) [reply src-extras/Database/LSMTree/Extras/ReferenceImpl.hs:97]: Nope. That's a copy'n'past error. Thanks!

File: prototypes/FormatPage.hs [hunk 9]
```diff
@@ -7,33 +7,52 @@
...
 -- | This accompanies the format-page.md documentation as a sanity check
 -- and a precise reference. It is intended to demonstrate that the page
--- format works.
+-- format works. It is also used as a reference implementation for tests of
+-- the real implementation.
+--
+-- Logically, a page is a sequence of key,operation pairs (with optional
+-- blobrefs), sorted by key, and its serialised form fits within a disk page.
+--
+-- The reference implementation does not rely on the keys being sorted.
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply prototypes/FormatPage.hs:16]: I've tweaked the wording:

> This reference implementation covers serialisation and deserialisation (not lookups) which do not rely on (or enforce) the keys being sorted.

**Comment by dcoutts** [comment]
Ok, I think I've addressed everything except for the test of the invariant of the sorted test cases.

File: prototypes/FormatPage.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jorisdral) [reply prototypes/FormatPage.hs:835]: After giving it a little more thought, I'm not sure I have a better suggestion right now

**Review by jorisdral — COMMENTED** [review]
LGTM. If you want to add the missing invariant test, feel free to include  it and merge

**Review by jorisdral — APPROVED** [review]

**Pull request merged**